R/board_ms365.R

```diff
@@ -0,0 +1,188 @@
+#' Use a OneDrive or Sharepoint document library as a board
+#'
+#' Pin data to a folder in Onedrive or a SharePoint Online document library
+#' using the Microsoft365R package.
+#'
+#' @inheritParams new_board
+#' @param drive A OneDrive or SharePoint document library object, as obtained
+#'   from Microsoft365.
+#' @param path Path to directory to store pins. Will be created if it
+#'   doesn't already exist.
+#' @param delete_by_item Whether to handle folder deletions on an item-by-item
+#'   basis, rather than deleting the entire folder at once. You may need to set
+#'   this to `TRUE` for a board in SharePoint Online or OneDrive for Business,
+#'   due to document protection policies that prohibit deleting non-empty
+#'   folders.
+#' @export
+#' @examples
+#' \dontrun{
+#' # A board in your personal OneDrive
+#' od <- Microsoft365R::get_personal_onedrive()
+#' board <- board_onedrive(od, "myboard")
+#' board %>% pin_write(iris)
+#'
+#' # A board in OneDrive for Business
+#' odb <- Microsoft365R::get_business_onedrive(tenant = "mytenant")
+#' board <- board_onedrive(odb, "myproject/board")
+#'
+#' # A board in a SharePoint Online document library
+#' sp <- Microsoft365R::get_sharepoint_site("my site", tenant = "mytenant")
+#' doclib <- sp$get_drive()
+#' board <- board_sharepoint(doclib, "general/project1/board")
+#' }
+board_ms365 <- function(drive, path, versioned = TRUE, cache = NULL, delete_by_item = FALSE) {
+  check_installed("Microsoft365R")
+
+  if (!inherits(drive, "ms_drive")) {
+    abort("`drive` must be a OneDrive or SharePoint document library object")
+  }
+
+  # try to create the board folder: ignore error if folder already exists
+  try(drive$create_folder(path), silent = TRUE)
+  folder <- drive$get_item(path)
+  if (!folder$is_folder()) {
+    abort("Invalid path specified")
+  }
+
+  cache <- cache %||% board_cache_path(paste0("ms365-", hash(folder$properties$id)))
+  new_board_v1("pins_board_ms365",
+    folder = folder,
+    cache = cache,
+    versioned = versioned,
+    delete_by_item = delete_by_item
+  )
+}
+
+#' @export
+#' @rdname board_ms365
+board_sharepoint <- board_ms365
+
+#' @export
+#' @rdname board_ms365
+board_onedrive <- board_ms365
+
+board_ms365_test <- function(...) {
+  skip_if_missing_envvars("board_ms365()", "PINS_MS365_TEST_DRIVE")
```

Comment:
How did you make this object? Is there some way I could extract data out of `Microsoft365R::get_personal_onedrive()` to get a re-usable token, or will this always require re-auth?  Your full "industrial solution" seems a bit heavy — is there something that's maybe a bit hacky but is lighter weight?

Reply:
It's pretty simple:

```r
od <- get_personal_onedrive()
saveRDS(od, "path/to/file.rds")
Sys.setenv(PINS_MS365_TEST_DRIVE="path/to/file.rds")
```

I assume you're testing S3 and Azure storage in the pipeline? You could copy the rds file into a bucket/container and download it from there as part of the pipeline. It's an R6 object and pretty much self-contained.

If you don't want to copy the full drive object, the OAuth token itself is in the `$token` component. It has a very similar structure to a `httr::Token` object. You can pass this to `get_personal_onedrive` to recreate the drive object:

```r
token <- od$token
saveRDS(token, "path/to/file.rds")

...

token_in_testing <- readRDS("file.rds")
od_in_testing <- get_personal_onedrive(token=token_in_testing)
```

The main caveat is that the embedded refresh token has a [default lifetime](https://docs.microsoft.com/en-us/azure/active-directory/develop/refresh-tokens) of 90 days. This means that as long as you're running the pipeline every 90 days, everything will work; after that, the token will run out and you'll have to reupload a new object. I'm also not entirely sure if it's kosher to be copying tokens from one machine to another, but nothing seems to complain.

Reply:
Ok, I've made a small hack to make interactive testing easier (set an env var to allow using your personal one drive) and a note to look into this again when I have more time (#482)